File: 2022/csharp/Day03.cs
namespace Aoc2022;


public class Day03 : Day
{
    static int Score(char c) => 1 + (char.IsUpper(c) ? (c + 26 - 'A') : c - 'a');

    public override string SolveA(string input)
    {
        var score = input.Split('\n').Select(line =>
        {
            var half = line.Length / 2;
            var (left, right) = (line[..half], line[half..]);
            return left.Intersect(right).First();
        })
        .Select(Score)
        .Sum();
        return score.ToString();
    }

    public override string SolveB(string input)
    {
        var groups = input.Split('\n').Chunk(3);
        var badges = groups.Select(group => group[0].Intersect(group[1]).Intersect(group[2]).First());
        return badges.Select(Score).Sum().ToString();
    }

    public Day03()
    {
        Tests = new()
        {
            new("Score", "A", "27", s => Score(s[0]).ToString()),
            new(
                "A",
                """
                vJrwpWtwJgWrhcsFMMfFFhFp
                jqHRNqRjqzjGDLGLrsFMfFZSrLrFZsSL
                PmmdzqPrVvPwwTWBwg
                wMqvLMZHhHMvwLHjbvcjnnSBnvTQFn
                ttgJtRGJQctTZtZT
                CrZsJsPPZsGzwwsLwLmpwMDw
                """,
                "157",
                SolveA),
            new(
                "B",
                """
                vJrwpWtwJgWrhcsFMMfFFhFp
                jqHRNqRjqzjGDLGLrsFMfFZSrLrFZsSL
                PmmdzqPrVvPwwTWBwg
                wMqvLMZHhHMvwLHjbvcjnnSBnvTQFn
                ttgJtRGJQctTZtZT
                CrZsJsPPZsGzwwsLwLmpwMDw
                """,
                "70",
                SolveB
            )
        };
    }
}

File: 2022/fsharp/Day03.fs
module Aoc.Solutions.Day03

open Aoc.Runner

let score (c: char) = 
    1 + 
    (if System.Char.IsUpper(c) then
        (int)c - (int)'A' + 26
    else (int)c - (int)'a')


type Day03() =
    inherit Day()

    override _.SolveA input =
        input.Split '\n'
        |> Seq.map (fun line ->
            let half = line.Length / 2
            let left = line[..(half - 1)]
            let right = line[half..]
            (Set.intersect (Set.ofSeq left) (Set.ofSeq right)) |> Seq.head)
        |> Seq.map score
        |> Seq.sum
        |> string


    override _.SolveB input =
        input.Split '\n'
        |> Seq.map Set.ofSeq
        |> Seq.chunkBySize 3
        |> Seq.map Set.intersectMany
        |> Seq.map Seq.head
        |> Seq.map score
        |> Seq.sum
        |> string